[clang][DebugInfo] Add virtuality call-site target information in DWARF.

CarlosAlbertoEnciso opened the pull request:
Given the test case:
```
struct CBase {
  virtual void foo();
};

void bar(CBase *Base) {
  Base->foo();
}
```
and using '-emit-call-site-info' with 'llc', currently the following DWARF is produced for the indirect call 'Base->foo()':
```
0x10: DW_TAG_structure_type
        DW_AT_name	("CBase")

0x20:   DW_TAG_subprogram
          DW_AT_linkage_name	("_ZN5CBase3fooEv")
          DW_AT_name	("foo")
          ...

0x30: DW_TAG_subprogram
        DW_AT_linkage_name	("_Z3barP5CBase")
        DW_AT_name	("bar")
        ...

0x40:   DW_TAG_call_site
          DW_AT_call_target_clobbered	(DW_OP_breg0 RAX+0)
          DW_AT_call_return_pc	(0x000000000000000a)
```
Initial context from the SCE debugger point of view:

> We can detect when a function has been affected by Identical Code Folding (ICF) from DWARF call-site information. For example,
> 
> 1. $RIP is currently in 'bar()'
> 2. The call-site information in the parent frame indicates we should have called 'foo()'
> 
> If we see a situation like the above, we can infer that 'foo()' and 'bar()' have been code folded. This technique breaks when dealing with virtual functions because the call-site information only provides a way to find the target function at runtime.
> 
> However, if the call-site information includes information on which virtual function is being called, we can compare this against the $RIP location to see if we are in an implementation of the virtual function. If we are not, then we can assume we have been code folded.
> 
> For this to work we just to need to record which virtual function we are calling. We do not need to know the type of the 'this' pointer at the call-site.
> 

This patch (available for all debuggers) helps in the identification of the intended target of a virtual call in the SCE debugger.

By adding the ``DW_AT_LLVM_virtual_call_origin`` for indirect calls, a debugger can identify the intended target of a call. These are the specific actions taking by the SCE debugger:

- The debugger can detect functions that have been folding by comparing whether the ``DW_AT_call_origin`` matches the call frame function. If it does not, the debugger can assume the "true" target and the actual target have been code folded and add a frame annotation to the call stack to indicate this. That, or there is a tail call from ``foo`` to ``bar``, but the debugger can disambiguate these cases by looking at the ``DW_AT_call_origin`` referenced subroutine DIE which has a tombstone ``DW_AT_low_pc`` in the ICF case.

- For virtual calls such as the given test case, the existence of the ``DW_AT_LLVM_virtual_call_origin`` attribute tells the debugger that this is an indirect jump, and the ``DW_AT_LLVM_virtual_call_origin`` attribute, pointing to the base class method DIE, will tell which method is being called.

- The debugger can confirm from the method's DIE that it is a virtual function call by looking at the attributes (``DW_AT_virtuality``, and ``DW_AT_vtable_elem_location``) and can look at the parent DIE to work out the type.

This is the added ``DW_AT_LLVM_virtual_call_origin`` to identify the target call ``CBase::foo``.
```
0x40:   DW_TAG_call_site
          DW_AT_call_target_clobbered	(DW_OP_breg0 RAX+0)
          DW_AT_call_return_pc	(0x000000000000000a)
          -----------------------------------------------
          DW_AT_LLVM_virtual_call_origin	(0x20 "_ZN5CBase3fooEv")
          -----------------------------------------------
```
The extra call site information is available by default for all debuggers.

CarlosAlbertoEnciso commented:
@tromey For some reasons, your name does not appear in the reviewers list. I would appreciate it you can add yourself as reviewer. Thanks

jryans reviewed (commented):
Overall, looking good to me with a few small things to fix mentioned inline. Thanks for working on this!

Since this touches on a few areas I am not as confident in, I think would be for one more person to also take a look.

tromey commented:
> @tromey For some reasons, your name does not appear in the reviewers list. I would appreciate if you can add yourself as reviewer. Thanks

I don't have any permissions in llvm and so I think I can't be added as a reviewer.

Anyway, I didn't want your request to go unanswered, but I don't actually know that much about the code that this patch touches.  My main question was why it was conditional on a particular debugger but that's been addressed.

CarlosAlbertoEnciso commented on llvm/lib/Target/X86/X86ISelLoweringCall.cpp:
```diff
@@ -2050,8 +2050,14 @@ X86TargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
+  // b) We want to annotate indirect calls.
+  if (CB && CB->isIndirectCall() &&
+      (MF.getTarget().Options.EmitCallGraphSection ||
+       MF.getTarget().Options.EmitCallSiteInfo))
```
Very good point. I think the best way is to include them in this PR.

CarlosAlbertoEnciso commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5958,6 +5958,9 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
     }
   }
 
+  if (CGDebugInfo *DI = CGM.getModuleDebugInfo())
```
Moving the change into that block it make sense, as they are related.

OCHyams commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -2430,6 +2430,9 @@ llvm::DISubprogram *CGDebugInfo::CreateCXXMemberFunction(
   SPCache[Method->getCanonicalDecl()].reset(SP);
 
+  // Add the method declaration as a call target.
+  addCallTarget(MethodLinkageName, SP, /*CI=*/nullptr);
```
Possibly dumb question: What is this call doing?

CarlosAlbertoEnciso replied:
There are 3 items that are required to have a complete entry for the call-site information:
``Function Name``, ``Metadata Node`` and ``Instruction``.
They are collected as a whole set or individually. Here we have only ``Function Name``, ``Metadata Node``. The ``Instruction`` was collected or will be collected, in order to have the full set.

OCHyams commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -5082,8 +5180,13 @@ void CGDebugInfo::EmitFunctionEnd(CGBuilderTy &Builder, llvm::Function *Fn) {
 
-  if (Fn && Fn->getSubprogram())
+  if (Fn && Fn->getSubprogram()) {
+    // For each call instruction emitted, add the call site target metadata.
```
nit: comment mentions call instruction, but we pass nullptr for the call here - can the comment be re-wording?

CarlosAlbertoEnciso replied:
Comment reworded:
```
    // As the debug info for the given function has been emitted, add its
    // name and node to the call site target information.
```

OCHyams commented on clang/test/DebugInfo/CXX/callsite-base.cpp:
```diff
@@ -0,0 +1,36 @@
+// CHECK-BASE-DAG:   call void %3{{.*}} !dbg {{![0-9]+}}, !call_target [[BASE_TWO:![0-9]+]]
+// CHECK-BASE-DAG:   call void %5{{.*}} !dbg {{![0-9]+}}, !call_target [[BASE_THREE:![0-9]+]]
+// CHECK-BASE-DAG:   call void @_ZN5CBaseC2Ev{{.*}} !dbg {{![0-9]+}}
+// CHECK-BASE-DAG:   call void @_ZN5CBase3oneEv{{.*}} !dbg {{![0-9]+}}
```
I don't think these need to be -DAG checks since the calls will always get emitted in this order? Same applies to the non-metadata checks in the other tests.

OCHyams reviewed (commented):
I feel that this is large enough that is would benefit reviewers to split into LLVM changes followed by Clang changes. If it's not too much work to reshuffle, I think that would be helpful, but I won't insist on it! I suppose doing so now risks losing comments already made.

Please can you explain the intended use case a bit more, both for others in future in the commit message, but also for me now as I'm a little confused -

In your example you said that:
```
int function(CBase *Input) {
  int Output = Input->foo(); // This call gets annotated with call_origin (CBase::foo)
  return Output;
}
```

And `CBase::foo` is virtual. DWARF says:

> The call site entry may have a DW_AT_call_origin attribute which is a reference. For direct calls or jumps where the called subprogram is known it is a reference to the called subprogram’s debugging information entry. For indirect calls it may be a reference to a DW_TAG_variable, DW_TAG_formal_parameter or DW_TAG_member entry representing the subroutine pointer that is called.

Is it not bending the spec a little to have `DW_AT_call_origin` be a subprogram reference for a virtual call (which is indirect)? Possibly naively, would it make more sense for this this point to the global `_vtable$` variable DIE? I might have completely the wrong end of the stick here.

dwblaikie commented:
Yeah, seems problematic if the DWARF for a devirtualized or non-virtual call (`derived->Base::func()` for instance) is indistinguishable from a virtual call. Is that what's being proposed?

Anyone have suggestions on how we can avoid that ^ property? Ideally without having consumers need to know about something extra to detect this case (eg: putting an extra extension attribute on the call site die that says "this is a virtual call" is probably insufficient - because consumers that don't understand the extension would be left in the "unable to differentiate" case) - so maybe requires an extension attribute to describe the target?

jmorse commented:
> Yeah, seems problematic if the DWARF for a devirtualized or non-virtual call (derived->Base::func() for instance) is indistinguishable from a virtual call. Is that what's being proposed?

The interpretation I've been taking here is that we're providing more information where there was an absence of it, rather than changing a meaning. As far as I'm aware, we haven't been distinguishing virtual/non-virtual call sites in the past, just not recording extra information about indirect calls. (85% confidence here -- I think we've been recording "the call target is this register" in call-site info, but nothing more?). This patch would be providing new information that's narrowly true about the source code, following the example that Orlando has pulled out, the source-code _is_ a call to CBase::foo, but it also happens to be indirect due to it being a virtual call.

I suppose my question is -- have consumers really been using the absence of a subprogram reference as a proxy for "this is a virtual call", or can we get away with not adding another attribute? If there's a serious risk that consumers are doing that, an attribute is fine; presumably in the number space we should place it immediately after DW_AT_LLVM_alloc_type?

CarlosAlbertoEnciso commented:
Given the test case:
```
class CBaseOne {
  virtual void foo(int &);
};

struct CDerivedOne : CBaseOne {
  void foo(int &);
};

void CDerivedOne::foo(int &) {
}

struct CBaseTwo {
  CDerivedOne *DerivedOne;
};

struct CDerivedTwo : CBaseTwo {
  void bar(int &);
};

void CDerivedTwo::bar(int &j) {
  DerivedOne->foo(j);
}
```
This is the generated DWARF
```
0x00000072:   DW_TAG_subprogram
                DW_AT_specification	(0x00000041 "_ZN11CDerivedOne3fooERi")
                ...

0x00000091:   DW_TAG_structure_type
                DW_AT_name	("CDerivedTwo")
                ...

0x0000009d:     DW_TAG_subprogram
                  DW_AT_linkage_name	("_ZN11CDerivedTwo3barERi")
                  DW_AT_name	("bar")

0x000000c8:   DW_TAG_subprogram
                ...
                DW_AT_specification	(0x0000009d "_ZN11CDerivedTwo3barERi")

0x000000ea:     DW_TAG_call_site
                  DW_AT_call_target	(DW_OP_reg0 RAX)
                  DW_AT_call_tail_call	(true)
                  DW_AT_call_pc	(0x0000000000000019)
                  DW_AT_call_origin	(0x00000072 "_ZN11CDerivedOne3fooERi")

```

CarlosAlbertoEnciso commented:
@dwblaikie Thanks very much for your great input.
After some discussions with our debugger team, adding a new attribute to eliminate the current ambiguity works well for them,
A ``DW_AT_LLVM_virtual_call_origin`` has been suggested.

OCHyams commented:
Ah there's a detail about our debugger that turns out I was a bit out of date on. At load time the debugger makes a note of function symbols with the same address as ICF functions. Stopped in a function, the debugger knows whether the function has been involved in ICF without looking at DWARF. But it does need the DWARF call site information to determine whether the current function was the original callee (ICF "survivor") or not (ICF folded-away). _That_ is how call site information is used in relation to ICF in our debugger.

> Rather than trying to detect tail calls - can't you rely the call_site to tell you if it's a tail call? 

Yes and no. The call site info in the parent frame describes the call to the function that does the tail call, not the tail-called function:

```
Call graph:
  a() --direct-call-> b() --tail-call-> c()

Stopped in c() our call stack is:
  c()
  a() - Call site: DW_AT_call_origin('b')
```

If we ignore analyising the call graph for a moment, there's no way to get at the call site info of the call to `c()` in `b()` to see the tail call attribute. But yes, if we analyse the call graph we can in some cases work it out, though not all (e.g. ambiguous paths).

> until you have a graph from caller to callee following only tail call edges and check that the callee appears nowhere else in that graph

It turns out that our debugger does actually do this kind of call graph analysis to reconstruct tail-called frames.

Adding virtual_call_origin lets the debugger determine whether there has been a tail call for virtual calls. It's not completely watertight, because it fails in some cases such as:
```
Call graph:
   a() --virtual-call-> b() --tail-call-> Base::b()

Call stack:
  Base::b()
  a() - Call site: DW_AT_virtual_call_origin('b')
```

Here the debugger is going to assume `Base::b` was the intended call target, and report no tail call frames. Our debugger folks are happy with the trade off for the coverage gained.

Both ICF and entry value reconstruction are relevant here but I think maybe we can think of them separetely. The parent-frame-call-site-info to current-frame mismatch can occur due to ICF or tail calls. But the debugger already knows whether the current frame has been involved in ICF (see top of this comment), and I think all of this comes down to trying to work out if there's been a tail call.

>> I wonder, does this only help you detect ICF specifically on the first step into a call? I can't picture how this would work if you paused execution in the middle of the code-folded function.
>  because you could differentiate between the start of the function you're in and the caller's call_site's call_target

Oh right, ok that makes sense. So it works the same way and runs into the same issues with tail calls as direct/static calls do. Right I think I'm grasping the trade off here (ignoring performance/code size). The NOP-sled implementation tells you which function in the ICF folded group was originally called, but doesn't work if there's a tail call to the current frame. The current approach using the sybmol table indicates there's been ICF but not which function was orignally called, unless the callee is the the IFC "survivor". That logic is independant from the current call stack and DWARF, so doesn't get confused by tail calls. Going further, maybe these are these composable?

I've sent us round in circle a bit here, sorry about that.

To sumarise:

* The debugger wants to understand whether there's been a tail call to the current function.
* It can construct a tail-call-path between the parent frame and current frame today unless there's any indirect calls.
* We're proposing adding DW_AT_virtual_call_origin so the tail call graph can be constructed (imperfectly) in the presence of virtual calls, which is probably going to be a significant portion of indirect calls in a C++ codebase.

The NOP-sleds may be composable with rather than a replacement for what we're doing. However, either way, I think we'd need to understand the size/performance costs to be able to make a call on their use, and don't think we're able to tackle that investigation right away.

dwblaikie commented:
> Ah there's a detail about our debugger that turns out I was a bit out of date on. At load time the debugger makes a note of function symbols with the same address as ICF functions. Stopped in a function, the debugger knows whether the function has been involved in ICF without looking at DWARF. But it does need the DWARF call site information to determine whether the current function was the original callee (ICF "survivor") or not (ICF folded-away). _That_ is how call site information is used in relation to ICF in our debugger.
> 
> > Rather than trying to detect tail calls - can't you rely the call_site to tell you if it's a tail call?
> 
> Yes and no. The call site info in the parent frame describes the call to the function that does the tail call, not the tail-called function:
> 
> ```
> Call graph:
>   a() --direct-call-> b() --tail-call-> c()
> 
> Stopped in c() our call stack is:
>   c()
>   a() - Call site: DW_AT_call_origin('b')
> ```
> 
> If we ignore analyising the call graph for a moment, there's no way to get at the call site info of the call to `c()` in `b()` to see the tail call attribute. But yes, if we analyse the call graph we can in some cases work it out, though not all (e.g. ambiguous paths).
> 
> > until you have a graph from caller to callee following only tail call edges and check that the callee appears nowhere else in that graph
> 
> It turns out that our debugger does actually do this kind of call graph analysis to reconstruct tail-called frames.
> 
> Adding virtual_call_origin lets the debugger determine whether there has been a tail call for virtual calls. It's not completely watertight, because it fails in some cases such as:
> 
> ```
> Call graph:
>    a() --virtual-call-> b() --tail-call-> Base::b()
> 
> Call stack:
>   Base::b()
>   a() - Call site: DW_AT_virtual_call_origin('b')
> ```
> 
> Here the debugger is going to assume `Base::b` was the intended call target, and report no tail call frames. Our debugger folks are happy with the trade off for the coverage gained.

Would `DW_AT_call_target` in `a()` tell you the real function being called (the address of `SomeDerived::b` - might have to jump through some thunks to get there) - and thus cover this case? (& others, without the need for `DW_AT_virtual_call_origin`)

> Both ICF and entry value reconstruction are relevant here but I think maybe we can think of them separetely. The parent-frame-call-site-info to current-frame mismatch can occur due to ICF or tail calls. But the debugger already knows whether the current frame has been involved in ICF (see top of this comment), and I think all of this comes down to trying to work out if there's been a tail call.
> 
> > > I wonder, does this only help you detect ICF specifically on the first step into a call? I can't picture how this would work if you paused execution in the middle of the code-folded function.
> > > because you could differentiate between the start of the function you're in and the caller's call_site's call_target
> 
> Oh right, ok that makes sense. So it works the same way and runs into the same issues with tail calls as direct/static calls do. Right I think I'm grasping the trade off here (ignoring performance/code size). The NOP-sled implementation tells you which function in the ICF folded group was originally called, but doesn't work if there's a tail call to the current frame. The current approach using the sybmol table indicates there's been ICF but not which function was orignally called, unless the callee is the the IFC "survivor". That logic is independant from the current call stack and DWARF, so doesn't get confused by tail calls. Going further, maybe these are these composable?
> 
> I've sent us round in circle a bit here, sorry about that.
> 
> To sumarise:
> 
> * The debugger wants to understand whether there's been a tail call to the current function.
> * It can construct a tail-call-path between the parent frame and current frame today unless there's any indirect calls.

Why can't it reconstruct this in the face of indirect calls - if those calls include `DW_AT_call_target` that resolves to the address of the 

> * We're proposing adding DW_AT_virtual_call_origin so the tail call graph can be constructed (imperfectly) in the presence of virtual calls, which is probably going to be a significant portion of indirect calls in a C++ codebase.

Perhaps this is the place we should focus - concrete examples of what is/isn't possible help me feel a bit more grounded in the conversation - sorry for any other tangents/misleading statements, etc.

So this is the `a() --virtual-call-> b() --tail-call-> Base::b()` example above? in this case, if `a` is non-tail-calling `SomeDerived::b`, the `DW_AT_call_target` should have `SomeDerived::b` in it - so you could use that to determine that `Base::b` wasn't the real target, and so there's been a tail call here somewhere (& with the call graph analysis, if `SomeDerived::b` only has one tail call chain that leads to `Base::b`, then you can reconstruct the call stack correctly - and if there's more than one such chain or possible chain (tail dynamic call sites, etc) then you bail out and say there could be missing frames here)

> The NOP-sleds may be composable with rather than a replacement for what we're doing. However, either way, I think we'd need to understand the size/performance costs to be able to make a call on their use, and don't think we're able to tackle that investigation right away.

Understandable - though I'm not sure I've quite understood the composability/orthogonality yet - but that's on me & probably best set aside for now - I'll take your word for it.

dwblaikie commented on clang/test/DebugInfo/CXX/callsite-derived.cpp:
```diff
@@ -0,0 +1,77 @@
+}
+
+class CDerived : public CBase {
+public:
```
probably use a `struct` (here and elsewhere) since it's public inheritance and members by default

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -4955,6 +4958,95 @@ void CGDebugInfo::EmitFunctionDecl(GlobalDecl GD, SourceLocation Loc,
+      InstrList &List = It->second.second;
+      List.push_back(CI);
+    }
+  }
```
This is doing some duplicate lookup - by using `find` then `emplace` it does the map lookup twice - usually good to try to avoid that, and it often simplifies the code too... I /think/ this is equivalent and simpler:
```suggestion
  auto It = CallTargetCache.insert(Name);
  if (MD)
    It->second.first.reset(MD);
  if (CI)
    It->second.second.push_back(CI);
```

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -4955,6 +4958,95 @@ void CGDebugInfo::EmitFunctionDecl(GlobalDecl GD, SourceLocation Loc,
+
+  auto It = CallTargetCache.find(Name);
+  if (It != CallTargetCache.end())
+    CallTargetCache.erase(It);
```
```suggestion
  CallTargetCache.erase(Name);
```

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -903,6 +912,25 @@ class CGDebugInfo {
+  ///   ('CallBase' or 'MDNode') is available.
+  using InstrList = llvm::SmallVector<llvm::WeakVH, 2>;
+  using CallTargetEntry = std::pair<llvm::TrackingMDNodeRef, InstrList>;
+  llvm::SmallDenseMap<StringRef, CallTargetEntry> CallTargetCache;
```
This infrastructure seems a bit complicated/unforttunate - can we avoid the "cache" and conditional supporting by supporting this in the same way we inject function declarations into classes as-needed when emitting debug info for out of line definitions. Consider this example: https://godbolt.org/z/qdM5d35hc - the function declaration is emitted (& will be deduplicated (or won't be duplicated in the first place, I forget how we do it) if later in the file a definition was emitted (eg: if the ctor definition came after the ::f2 definition)) - we could always produce a function declaration and a class declaration for the call site - the metadata in the call_target would keep both alive, unless the call site were optimized away, in which case the type and declaration would be optimized away.

Doing this might also allow us a better representation even for the non-virtual cases, where I believe we have similar problems where the call site debug info is conditionally emitted only if the function declaration is emitted for some other reason - ie: the call site doesn't keep the debug info alive as it should. Which is awkward for users since the reasons the call site might appear or not are pretty opaque/unrelated to what the user is doing.

CarlosAlbertoEnciso replied:
The `CallTargetCache` map has been removed.

dwblaikie commented on clang/test/DebugInfo/CXX/callsite-edges.cpp:
```diff
@@ -0,0 +1,112 @@
+// The following are identified edge cases involving the method being called:
+// 1) Method is declared but not defined in current CU.
+// 2) Pure virtual method but not defined in current CU.
+// 3) Virtual method defined very deeply.
```
"very deeply" seems a bit vague - could you be more specific about what conditions are a problem/why they might be a problem?

dwblaikie commented on clang/test/DebugInfo/CXX/callsite-edges.cpp:
```diff
@@ -0,0 +1,112 @@
+//---------------------------------------------------------------------
+// 1) Method is declared but not defined in current CU - Fail.
+//    No debug information entry is generated for 'one'.
+//    Generate 'call_target' metadata only for 'two'.
```
Hopefully my suggested direction elsewhere (re: lazily producing the function declarations) should address this?

dwblaikie commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -9418,8 +9418,14 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
+  // b) We want to annotate indirect calls.
+  if (CB && CB->isIndirectCall() &&
+      (MF.getTarget().Options.EmitCallGraphSection ||
+       MF.getTarget().Options.EmitCallSiteInfo))
```
This condition is appearing several times - can it be tested for in fewer/one place and have the consequences flow out more naturally/with fewer special cases spread around? Or, if that's not possible, at least the condition may need to be refactored into a reusable function.

dwblaikie reviewed (commented):
Some thoughts

dwblaikie commented:
> The tangents are all good and I am also responsible there. All these features are connected and I've found it difficult to disentangle the parts doing interesting things from other stuff that's just nearby. Thanks for working it through and bearing with me/us.
> 
> > Why can't it reconstruct this in the face of indirect calls - if those calls include DW_AT_call_target that resolves to the address of the [called function]
> 
> Ah, I think you're right that would also work. I think essentially it's a different way to achieve the same the result. The call_target approach looks like it has much poorer coverage though as it can't be used if the call_target expressions contain volatile registers.
> 
> I've got a patch to emit `DW_AT_call_target_clobbered` when volatile regs are used in the expr that I'm sitting on but wanted to wait for this stuff to get resolved. Using that patch (on top of another that allows memory locations in call_target expressions), Using llvm-dwarfdump on a self-host RelWithDebInfo Clang build I count 113899 instances of `DW_AT_call_target_clobbered` versus only 2229 instances of `DW_AT_call_target` in the DWARF.

Thanks for gathering the data - fair enough.

You folks are the ones mostly driving the call_site work, so that's probably about enough for me to be OK with this extension attribute for your own experimentation.

CarlosAlbertoEnciso committed:
[clang][DebugInfo] Add call site target information in DWARF.

These changes are intended to address the concern about
the 'CallTargetCache' map.

The map is removed by:
- injecting a function declaration if one is not available
  at the point where the function is invoked.
- always pointing the 'call_target' to the function declaraion.

CarlosAlbertoEnciso commented:
The updated patch includes:
- lazily producing the function declarations (suggested by @dwblaikie)
- the DW_AT_LLVM_virtual_call_origin always pointing to the method declaration

dwblaikie reviewed (approved):
Haven't looked super close, but sounds roughly right to me

CarlosAlbertoEnciso commented on clang/test/DebugInfo/CXX/callsite-edges.cpp:
```diff
@@ -0,0 +1,117 @@
+struct CDeep {
+  struct CD1 {
+    struct CD2 {
+      struct CD3 {
```
The depth was relevant for the previous implementation (using a map).
Reduced the test case, as a nesting level of 2 is just fine.

CarlosAlbertoEnciso committed:
[clang][DebugInfo] Add call site target information in DWARF.

Address reviewer feedback:
- Rename 'MD' variable to a better descriptive name.
- Remove redundant code for indirect call check.
- Reduce nesting level in 'callsite-edges' test case.
- Improve comments and add check for CBase type definition.

CarlosAlbertoEnciso committed:
[clang][DebugInfo] Add virtuality call-site target information in DWARF.

Given the test case:

  struct CBase {
    virtual void foo();
  };

  void bar(CBase *Base) {
    Base->foo();
  }

and using '-emit-call-site-info' with llc, the following DWARF
is produced for the indirect call 'Base->foo()':

1$: DW_TAG_structure_type "CBase"
      ...
2$:   DW_TAG_subprogram "foo"
        ...

3$: DW_TAG_subprogram "bar"
      ...
4$:   DW_TAG_call_site
        ...

We add DW_AT_LLVM_virtual_call_origin to existing call-site
information, linking indirect calls to the function-declaration
they correspond to.

4$:   DW_TAG_call_site
        ...
        DW_AT_LLVM_virtual_call_origin (2$ "_ZN5CBase3fooEv")

The new attribute DW_AT_LLVM_virtual_call_origin helps to
address the ambiguity to any consumer due to the usage of
DW_AT_call_origin.

The functionality is available to all supported debuggers.

felipepiovezan commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -682,6 +682,9 @@ class CGDebugInfo {
   llvm::DINode::DIFlags getCallSiteRelatedAttrs() const;
 
+  /// Add call target information.
+  void addCallTarget(const FunctionDecl *FD, llvm::CallBase *CI);
```
since this only adds _virtual_ call targets, should it be called `addCallTargetIfVirtual`?

felipepiovezan commented on cross-project-tests/debuginfo-tests/clang_llvm_roundtrip/callsite-dwarf.cpp:
```diff
@@ -0,0 +1,68 @@
+// RUN: %clang --target=x86_64-unknown-linux -c -g -O1 %s -o - | \
```
Yeah I'd be more interested in seeing a test case later inside LLDB to make use of this feature.

felipepiovezan commented on llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp:
```diff
@@ -1891,3 +1887,12 @@ DIE *DwarfCompileUnit::getOrCreateSubprogramDIE(const DISubprogram *SP,
 }
+
+void DwarfCompileUnit::addLinkageNamesToDeclarations(
+    const DwarfDebug *DD, const DISubprogram *CalleeSP, DIE *CalleeDIE) {
```
If we're going to unconditionally dereference pointers, we should make them references instead to express our intent that this function does not know how to handle nullptrs

felipepiovezan reviewed (approved):
LGTM, left some comments about style that should be straightforward to address

CarlosAlbertoEnciso commented on llvm/lib/CodeGen/MIRPrinter.cpp:
```diff
@@ -541,7 +541,7 @@ static void convertCallSiteObjects(yaml::MachineFunction &YMF,
     YmlCS.CallLocation = CallLocation;
 
-    auto [ArgRegPairs, CalleeTypeIds] = CallSiteInfo;
+    auto [ArgRegPairs, CalleeTypeIds, MD] = CallSiteInfo;
```
Changed to `_`.

CarlosAlbertoEnciso committed:
[clang][DebugInfo] Add virtuality call-site target information in DWARF.

Address additional reviewer feedback:
- Rename a couple of functions to add clarity.
- Remove some comments that don't add anything.
- Change function signature to use references.
